From bbac3cacff9db9924aaba22bcefa3912024b1451 Mon Sep 17 00:00:00 2001
From: Manolo van Ee <manolo.vanee@gmail.com>
Date: Thu, 11 Dec 2025 22:01:02 +0100
Subject: [PATCH 2/4] [stdlib Concurrency] Update Clock.cpp for use with newlib

- Small fixes to make the code more portable.
- Use _POSIX_TIMERS to select use of nanosleep for swift_sleep.
---
 stdlib/public/Concurrency/Clock.cpp | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/stdlib/public/Concurrency/Clock.cpp b/stdlib/public/Concurrency/Clock.cpp
index 0ae357a5dda..f98af356ef1 100644
--- a/stdlib/public/Concurrency/Clock.cpp
+++ b/stdlib/public/Concurrency/Clock.cpp
@@ -62,12 +62,12 @@ void swift_get_time(
       continuous.tv_nsec = (interruptTime % 10'000'000) * 100;
 #elif WE_HAVE_STD_CHRONO
       auto now = std::chrono::steady_clock::now();
-      auto epoch = std::chrono::steady_clock::min();
+      auto epoch = std::chrono::steady_clock::time_point::min();
       auto timeSinceEpoch = now - epoch;
       auto sec = std::chrono::duration_cast<std::chrono::seconds>(timeSinceEpoch);
       auto ns = std::chrono::duration_cast<std::chrono::nanoseconds>(timeSinceEpoch - sec);
-      continuous.tv_sec = sec;
-      continuous.tv_nsec = ns;
+      continuous.tv_sec = sec.count();
+      continuous.tv_nsec = ns.count();
 #else
 #error Missing platform continuous time definition
 #endif
@@ -97,12 +97,12 @@ void swift_get_time(
       suspending.tv_nsec = (unbiasedTime % 10'000'000) * 100;
 #elif WE_HAVE_STD_CHRONO
       auto now = std::chrono::steady_clock::now();
-      auto epoch = std::chrono::steady_clock::min();
+      auto epoch = std::chrono::steady_clock::time_point::min();
       auto timeSinceEpoch = now - epoch;
       auto sec = std::chrono::duration_cast<std::chrono::seconds>(timeSinceEpoch);
       auto ns = std::chrono::duration_cast<std::chrono::nanoseconds>(timeSinceEpoch - sec);
-      suspending.tv_sec = sec;
-      suspending.tv_nsec = ns;
+      suspending.tv_sec = sec.count();
+      suspending.tv_nsec = ns.count();
 #else
 #error Missing platform suspending time definition
 #endif
@@ -137,7 +137,7 @@ switch (clock_id) {
       auto num = std::chrono::steady_clock::period::num;
       auto den = std::chrono::steady_clock::period::den;
       continuous.tv_sec = num / den;
-      continuous.tv_nsec = (num * 1000000000ll) % den
+      continuous.tv_nsec = (num * 1000000000ll) % den;
 #else
 #error Missing platform continuous time definition
 #endif
@@ -161,8 +161,8 @@ switch (clock_id) {
 #elif WE_HAVE_STD_CHRONO
       auto num = std::chrono::steady_clock::period::num;
       auto den = std::chrono::steady_clock::period::den;
-      continuous.tv_sec = num / den;
-      continuous.tv_nsec = (num * 1'000'000'000ll) % den
+      suspending.tv_sec = num / den;
+      suspending.tv_nsec = (num * 1'000'000'000ll) % den;
 #else
 #error Missing platform suspending time definition
 #endif
@@ -199,7 +199,7 @@ void swift_sleep(
     delay = deadline - now;
   }
 #elif defined(__linux__) || defined(__APPLE__) || defined(__wasi__) \
-  || defined(__OpenBSD) || defined(__FreeBSD__)
+  || defined(__OpenBSD) || defined(__FreeBSD__) || defined(_POSIX_TIMERS)
   struct timespec ts;
   ts.tv_sec = seconds;
   ts.tv_nsec = nanoseconds;
-- 
2.51.0

